libgraphics/src/drawing.cpp

```diff
@@ -0,0 +1,1894 @@
+                            t_opcode == kMCGDrawingPathOpcodeRelativeReflexArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeRelativeReverseArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeRelativeReverseReflexArcTo;
+                bool t_is_reflex =
+                            t_opcode == kMCGDrawingPathOpcodeReflexArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeRelativeReflexArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeReverseReflexArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeRelativeReverseReflexArcTo;
+                bool t_is_reverse =
+                            t_opcode == kMCGDrawingPathOpcodeReverseArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeRelativeReverseArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeReverseReflexArcTo ||
+                            t_opcode == kMCGDrawingPathOpcodeRelativeReverseReflexArcTo;
+                p_visitor.PathArcTo(t_is_relative, t_is_reflex, t_is_reverse, t_rx, t_ry, t_rotation, t_x, t_y);
+            }
+            break;
+                
+            case kMCGDrawingPathOpcodeBearing:
+            {
+                MCGFloat t_angle;
+                if (!AngleScalar(t_angle))
+                {
+                    break;
+                }
+                p_visitor.PathBearing(t_angle);
+            }
+            break;
+            
+            case kMCGDrawingPathOpcodeCloseSubpath:
+            {
+                p_visitor.PathCloseSubpath();
+            }
+            break;
+        }
+    }
+}
+
+/******************************************************************************/
+
+struct MCGDrawingRenderVisitor
+{
+    MCGContextRef gcontext = nullptr;
+
+    /**/
+    
+    bool paint_is_for_stroke = false;
+    bool paint_set = false;
+    MCGAffineTransform current_transform;
+    
+    MCGPoint first_point = {0, 0};
+    MCGPoint last_point = {0, 0};
+    MCGPoint last_control_point = {0, 0};
+    MCGDrawingPathOpcode last_curve_opcode = kMCGDrawingPathOpcodeEnd;
+    
+    /**/
+    
+    MCGPoint Point(bool p_is_relative, MCGFloat p_x, MCGFloat p_y)
+    {
+        if (p_is_relative)
+        {
+            return MCGPointMake(last_point.x + p_x, last_point.y + p_y);
+        }
+        return MCGPointMake(p_x, p_y);
+    }
+    
+    MCGPoint HorizontalPoint(bool p_is_relative, MCGFloat p_x)
+    {
+        if (p_is_relative)
+        {
+            return MCGPointMake(last_point.x + p_x, last_point.y);
+        }
+        return MCGPointMake(last_point.x + p_x, last_point.y);
+    }
+    
+    MCGPoint VerticalPoint(bool p_is_relative, MCGFloat p_y)
+    {
+        if (p_is_relative)
+        {
+            return MCGPointMake(last_point.x, last_point.y + p_y);
+        }
+        return MCGPointMake(last_point.x, last_point.y + p_y);
```

Comment:
I'm a bit confused by `H` versus `h` and `V` versus `v` - is it just in this spec that they're identical? It's annoying that that isn't explicit in the spec.

Reply:
That's an error - H is absolute, h is relative - the non-relative path should use p_x / p_y not last_point + p_x/p_y.